RecoParticleFlow/PFProducer/plugins/MLPFProducerSonic.cc

```diff
@@ -0,0 +1,199 @@
+    for (const auto& block : blocks) {
+      const auto& elems = block.elements();
+      for (const auto& elem : elems) {
+        if (all_elements.size() < NUM_MAX_ELEMENTS_BATCH) {
+          all_elements.push_back(&elem);
+        } else {
+          //model needs to be created with a bigger LSH codebook size
+          edm::LogError("MLPFProducer") << "too many input PFElements for predefined model size: " << elems.size();
+          break;
+        }
+      }
+    }
+    const long long int num_elements_total = all_elements.size();
+
+    //tensor size must be a multiple of the bin size and larger than the number of elements
+    const long long int tensor_size = LSH_BIN_SIZE * (num_elements_total / LSH_BIN_SIZE + 1);
+    assert(tensor_size <= NUM_MAX_ELEMENTS_BATCH);
+
+    auto& input1 = iInput.at("x");
+
+    //we ignore Sonic/Triton batching, as it doesn't create a dim-3 input for batch size 1.
+    //instead, we specify the batch dim as a model dim.
+    input1.setShape(0, 1);
+    input1.setShape(1, tensor_size);
+
+    auto data1 = std::make_shared<TritonInput<float>>(1);
+    auto& vdata1 = (*data1)[0];
+    vdata1.reserve(input1.sizeShape());
+
+    //Fill the input tensor
+    unsigned int ielem = 0;
+    for (const auto* pelem : all_elements) {
+      //get the PFElement as a pointer
+      const auto& elem = *pelem;
+
+      //prepare the input array from the PFElement
+      const auto& props = get_element_properties(elem);
+
+      //copy features to the input array
+      for (unsigned int iprop = 0; iprop < NUM_ELEMENT_FEATURES; iprop++) {
+        vdata1.push_back(normalize(props[iprop]));
+      }
+      ielem += 1;
+    }
+    input1.toServer(data1);
+  }
+
+  void produce(edm::Event& iEvent, edm::EventSetup const& iSetup, Output const& iOutput) override {
+    using namespace reco::mlpf;
+
+    std::vector<reco::PFCandidate> pOutputCandidateCollection;
+    const auto& output1 = iOutput.at("Identity");
+
+    //get the data of the first (and only) batch
+    const auto& out_data = output1.template fromServer<float>();
+
+    const auto num_elem = output1.shape()[1];
+
+    unsigned int num_pred_particles = 0;
+    for (unsigned int ielem = 0; ielem < num_elem; ielem++) {
+      //get the coefficients in the output corresponding to the class probabilities (raw logits)
+      std::vector<float> pred_id_logits;
+      for (unsigned int idx_id = 0; idx_id <= NUM_CLASS; idx_id++) {
+        pred_id_logits.push_back(out_data[0][ielem * NUM_OUTPUTS + idx_id]);
+      }
+
+      //get the most probable class PDGID
+      int pred_pid = pdgid_encoding[arg_max(pred_id_logits)];
+
+      //get the predicted momentum components
+      float pred_eta = out_data[0][ielem * NUM_OUTPUTS + IDX_ETA];
+      float pred_phi = out_data[0][ielem * NUM_OUTPUTS + IDX_PHI];
+
+      //put phi into a modular range
+      pred_phi = TMath::ATan2(TMath::Sin(pred_phi), TMath::Cos(pred_phi));
```

Comment:
prefer `cmath` to `TMath`

Reply:
actually, just use the proper function from https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaPhi.h